internal/service/streamprivatelinkendpoint/resource.go

```diff
@@ -27,125 +39,137 @@ type rs struct {
 	conversion.UpdateSchemaDescription(&resp.Schema)
 }
 
 func (r *rs) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
-	var tfModel TFModel
-	resp.Diagnostics.Append(req.Plan.Get(ctx, &tfModel)...)
+	var plan TFModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	// streamPrivatelinkEndpointReq, diags := NewAtlasReq(ctx, &tfModel)
-	// if diags.HasError() {
-	// 	resp.Diagnostics.Append(diags...)
-	// 	return
-	// }
+	streamPrivatelinkEndpointReq, diags := NewAtlasReq(ctx, &plan)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
 
-	// TODO: make POST request to Atlas API and handle error in response
+	projectID := plan.ProjectId.ValueString()
 
-	// connV2 := r.Client.AtlasV2
-	// if err != nil {
-	//	resp.Diagnostics.AddError("error creating resource", err.Error())
-	//	return
-	//}
+	connV2 := r.Client.AtlasV2
+	streamsPrivateLinkConnection, _, err := connV2.StreamsApi.CreatePrivateLinkConnection(ctx, projectID, streamPrivatelinkEndpointReq).Execute()
+	if err != nil {
+		resp.Diagnostics.AddError("error creating resource", err.Error())
+		return
+	}
 
-	// TODO: process response into new terraform state
-	// newStreamPrivatelinkEndpointModel, diags := NewTFModel(ctx, apiResp)
-	// if diags.HasError() {
-	// 	resp.Diagnostics.Append(diags...)
-	// 	return
-	// }
-	// resp.Diagnostics.Append(resp.State.Set(ctx, newStreamPrivatelinkEndpointModel)...)
-}
+	finalResp, err := waitStateTransition(ctx, projectID, *streamsPrivateLinkConnection.Id, connV2.StreamsApi)
+	if err != nil {
+		resp.Diagnostics.AddError("error when waiting for status transition in creation", err.Error())
+		return
+	}
 
-func (r *rs) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
-	var streamPrivatelinkEndpointState TFModel
-	resp.Diagnostics.Append(req.State.Get(ctx, &streamPrivatelinkEndpointState)...)
-	if resp.Diagnostics.HasError() {
+	newStreamPrivatelinkEndpointModel, diags := NewTFModel(ctx, projectID, finalResp)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
 		return
 	}
 
-	// TODO: make get request to resource
-
-	// connV2 := r.Client.AtlasV2
-	// if err != nil {
-	//	if apiResp != nil && apiResp.StatusCode == http.StatusNotFound {
-	//		resp.State.RemoveResource(ctx)
-	//		return
-	//	}
-	//	resp.Diagnostics.AddError("error fetching resource", err.Error())
-	//	return
-	//}
-
-	// TODO: process response into new terraform state
-	// newStreamPrivatelinkEndpointModel, diags := NewTFModel(ctx, apiResp)
-	// if diags.HasError() {
-	// 	resp.Diagnostics.Append(diags...)
-	// 	return
-	// }
-	// resp.Diagnostics.Append(resp.State.Set(ctx, newStreamPrivatelinkEndpointModel)...)
+	if plan.DnsSubDomain.IsNull() && len(newStreamPrivatelinkEndpointModel.DnsSubDomain.Elements()) == 0 {
+		newStreamPrivatelinkEndpointModel.DnsSubDomain = types.ListNull(types.StringType)
+	}
```

Comment:
q: Do we need to rely on the plan value? Could we potentially always store null if api response is empty?

Reply:
API returns an empty list instead of null when there are no dnsSubDomains defined, causing the following error in the provider: 
```
Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to mongodbatlas_stream_privatelink_endpoint.test, provider "provider[\"registry.terraform.io/mongodb/mongodbatlas\"]" produced an unexpected new
│ value: .dns_sub_domain: was null, but now cty.ListValEmpty(cty.String).
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```
this logic I added is similar to the `tags` attribute in other resources where we handle equally the null and empty list. Let me know if it makes sense

Reply:
Makes sense. I was more curious what are the cons of not relying on the `plan.DnsSubDomain` value, so if apiResp is nil or empty array we store nil in the state. The benefit I see from this is we remove this custom logic that lives outside of conversion logic and relies on the terraform plan value.

Reply:
if we rely solely in the api reponse and we store nil in the state when apiResp is nil or empty array there will be inconsistent results, for example, if we set `dns_sub_domain = []`, and we set nil in the state, this will happen:
```
Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to mongodbatlas_stream_privatelink_endpoint.test, provider "provider[\"registry.terraform.io/mongodb/mongodbatlas\"]" produced an unexpected new
│ value: .dns_sub_domain: was cty.ListValEmpty(cty.String), but now null.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

Reply:
I see, I guess we will have to continue supporting this additional logic. Only other alternative is that the API is capable for returning nil or [] depending on what is sent in the request, but I doubt this is feasible.